api/layer/object_cache_test.go

```diff
@@ -0,0 +1,170 @@
+
+	return id
+}
+
+func randSHA256Checksum(t *testing.T) (cs [sha256.Size]byte) {
+	_, err := rand.Read(cs[:])
+	require.NoError(t, err)
+
+	return
+}
+
+func TestTrimAfterObjectName(t *testing.T) {
+	var (
+		objects []*ObjectInfo
+		names   = []string{"b", "c", "d"}
+	)
+	for _, name := range names {
+		objects = append(objects, &ObjectInfo{Name: name})
+	}
+
+	t.Run("startafter before all objects", func(t *testing.T) {
+		actual := trimAfterObjectName("a", objects)
+		require.Equal(t, objects, actual)
+	})
+
+	t.Run("startafter first object", func(t *testing.T) {
+		actual := trimAfterObjectName(names[0], objects)
+		require.Equal(t, objects[1:], actual)
+	})
+
+	t.Run("startafter second-to-last object", func(t *testing.T) {
+		actual := trimAfterObjectName(names[len(names)-2], objects)
+		require.Equal(t, objects[len(objects)-1:], actual)
+	})
+
+	t.Run("startafter last object", func(t *testing.T) {
+		actual := trimAfterObjectName(names[len(names)-1], objects)
+		require.Empty(t, actual)
+	})
+
+	t.Run("startafter after all objects", func(t *testing.T) {
+		actual := trimAfterObjectName("z", objects)
+		require.Nil(t, actual)
+	})
+
+	t.Run("empty objects", func(t *testing.T) {
+		actual := trimAfterObjectName(names[0], []*ObjectInfo{})
+		require.Nil(t, actual)
+	})
+
+	t.Run("nil objects", func(t *testing.T) {
+		actual := trimAfterObjectName(names[0], nil)
+		require.Nil(t, actual)
+	})
+
+	t.Run("empty startafter", func(t *testing.T) {
+		actual := trimAfterObjectName("", objects)
+		require.Equal(t, actual, objects)
+	})
+}
+
+func TestTrimAfterObjectID(t *testing.T) {
+	var (
+		objects     []*ObjectInfo
+		ids         []*object.ID
+		numberOfIDS = 3
+	)
+
+	for i := 0; i < numberOfIDS; i++ {
+		id := randID(t)
+		objects = append(objects, &ObjectInfo{id: id})
+		ids = append(ids, id)
+	}
+
+	t.Run("existing id", func(t *testing.T) {
+		actual := trimAfterObjectID(ids[0].String(), objects)
+		require.Equal(t, objects[1:], actual)
+	})
+
+	t.Run("second to last id", func(t *testing.T) {
+		actual := trimAfterObjectID(ids[len(ids)-2].String(), objects)
+		require.Equal(t, objects[len(objects)-1:], actual)
+	})
+
+	t.Run("non-existing id", func(t *testing.T) {
+		actual := trimAfterObjectID("z", objects)
+		require.Nil(t, actual)
+	})
+
+	t.Run("last id", func(t *testing.T) {
+		actual := trimAfterObjectID(ids[len(ids)-1].String(), objects)
+		require.Nil(t, actual)
+	})
+
+	t.Run("empty id", func(t *testing.T) {
+		actual := trimAfterObjectID("", objects)
+		require.Nil(t, actual)
+	})
+}
+
+func TestObjectsListV2Cache(t *testing.T) {
+	var (
+		cacheSize = 10
+		objects   []*ObjectInfo
+		key       = "key"
+	)
+
+	for i := 0; i < cacheSize; i++ {
+		id := randID(t)
+		objects = append(objects, &ObjectInfo{id: id, Name: id.String()})
+	}
+
+	t.Run("lifetime", func(t *testing.T) {
+		cache := newListObjectsCache()
+		cache.Put(key, objects)
+
+		time.Sleep(defaultCacheLifetime - 10)
+		_, ok := cache.caches[key]
```

Comment:
We should not use cache directly. Better to use, for example, `cache.GetByContinuationToken(objects[0].ID().String(), key)`.

Reply:
I'm not sure because here we test only that cache keeps an entry for certain time. Methods GetBy... are tested below.  
At first, if we add GetBy... here,  the test will include another test for GetBy..., what seems redundant a bit.  
At second,  if I add GetBy... here and this test fail, it will not be clear  enough what is broken: cache's timer or GetBy...

Reply:
Otherwise tests run with a race detector will find a race.

Reply:
Fixed